### ISSUE TITLE
Type annotations?

### ISSUE THREAD

**Issue by evinism** [intake/filesystem_spec#625]
Are you interested in type annotations? E.g. if I added a bunch of type annotations to the project, is that something that you'd see being easily merged?

**Comment by martindurant** [comment]
Personally, I wouldn't say that I am "interested", but I am also not opposed. Call that ambivalent maybe.
I suspect it may be a great deal of work, and I am uncertain if, in the long run, it would entail more or less work for further development of the library. I am prepared to be convinced, and perhaps more voices should speak up here.

**Comment by martindurant** [comment]
cc @ianthomas23

**Comment by ianthomas23** [comment]
Thanks @aucampia, you present a strong case and a potential roadmap for starting this work. It is also evident that many people would like type annotations. No-one should be dissuaded by any lack of interest shown here, if a PR was submitted starting this work with some commitment to continue it, I would expect it to be merged without too much hassle. Or to put it another way, given that the python ecosystem is going in this direction I think it is inevitable that we will have to adopt type annotations eventually, it is just a question of when.

The problem really is that even if a large number of people request this then it makes no difference, what is needed is someone to actually do the work. It isn't a priority for established contributors, most of whom are inclined to fix problems that are impeding them, or add new functionality that they need. Maintainers with a wider view of the project are mostly working on consistency and reliability across the code base. Infrastructure issues like this often do not get near enough to the top of anybody's "todo" list to actually be actioned. There is a space here for a new contributor with type annotation experience to fill and if that doesn't happen probably an existing maintainer will eventually get around to it but I wouldn't expect it to be soon.

**Comment by aucampia** [comment]
> The problem really is that even if a large number of people request this then it makes no difference, what is needed is someone to actually do the work.

I agree with this and I understand it, it is significantly important to have fsspec be useful and defect free. I hope someone like @evinism finds the time to make this contribution. And hopefully whoever does have the time to do this will take care to do it with incrementally with minimal disruption.

**Comment by edgarrmondragon** [comment]
Would it be alright to get a list of modules, then as a first step configure mypy to ignore all of them and start tackling individually in separate PRs. I'd be happy to help with that as much as possible, but I wonder if that's too big of a burden on the maintainers.

**Comment by martindurant** [comment]
Having a mypy config which ignores all files would be fine with me, if it allows others a more obvious route to adding types to the codebase.

I believe there is a lot of duck-typing in this package, so I am not sure how easy it will be to get full typing coverage, but adding annotations to obvious places is probably a useful step. ([example](https://github.com/fsspec/filesystem_spec/issues/1669#issuecomment-2338185854) in the last five minutes)

**Comment by scholtalbers** [comment]
I'm starting out using this package with UPath and liking it so far! However, I do stumble on what arguments some methods expect. The draft MR #1676 would already gives me some quick hints, for example for `copy()`. 

I was doing `dest.fs.copy(src, dest, overwrite=overwrite, recursive=src.is_dir())` which fails, as both dest and src are UPath instances. Seeing that a str or a list of strings is expected would clear this up quickly in my editor (even if I would still expect UPath to be a supported argument.. :) )

```python
def copy(
        self,
        path1: str | list[str],
        path2: str | list[str],
        recursive: bool = False,
        maxdepth: int | None = None,
        on_error: Literal["ignore", "raise"] | None = None,
        **kwargs,
    ) -> None:
```

**Comment by martindurant** [comment]
Unfortunately, the signatures of methods have become increasingly complex with time. This is clearly poor style, but here we are - a bit like POSIX CLI commands.

For your specific case, Upath is built on fsspec, not the other way around. There is probably a way to do the copy operation as a method on the upath object itself, rather than expecting fsspec to check for a bunch of different types.

**Comment by slmnsh** [comment]
@martindurant I am working on this issue. I am going to convert `AbstractFileSystem` into `typing.Protocol` class and define multiple overloads for methods that accepts `list[str] | str`. Here is an example of overloading:

```python
@overload
@classmethod
def _strip_protocol(cls, path: str | os.PathLike[str] | Path) -> str: ...

@overload
@classmethod
def _strip_protocol(
    cls, path: list[str | os.PathLike[str] | Path]
) -> list[str]: ...

@classmethod
def _strip_protocol(
    cls, path: str | os.PathLike[str] | Path | list[str | os.PathLike[str] | Path]
) -> str | list[str]:
    """Turn path from fully-qualified to file-system-specific

    May require FS-specific handling, e.g., for relative paths or links.
    """
    if isinstance(path, list):
        return [cls._strip_protocol(p) for p in path]
    path = stringify_path(path)
    protos = (cls.protocol,) if isinstance(cls.protocol, str) else cls.protocol
    for protocol in protos:
        if path.startswith(protocol + "://"):
            path = path[len(protocol) + 3 :]
        elif path.startswith(protocol + "::"):
            path = path[len(protocol) + 2 :]
    path = path.rstrip("/")
    # use of root_marker to make minimum required path, e.g., "/"
    return path or cls.root_marker
```

It's time to modernize fsspec with proper typehints!

**Comment by martindurant** [comment]
I'll have a look at what you've done in the PR, but oh! how ugly the overloads are. I'm not certain How much information they are conveying for this particular function. You will probably have much more success with the less dispatch-oriented parts of the code, but there are many exceptions.

> convert AbstractFileSystem into typing.Protocol class

Is this compatible with the metaclass/caching strategy?

**Comment by slmnsh** [comment]
> but oh! how ugly the overloads are.

Yeah it's very useful for typehints, here are example of `ls`:

<img width="428" height="123" alt="Image" src="https://github.com/user-attachments/assets/1ca1931d-9447-420b-a86e-161c2b861958" />

<img width="486" height="125" alt="Image" src="https://github.com/user-attachments/assets/b9ceeda4-e4b2-451d-a700-979aba038208" />

> Is this compatible with the metaclass/caching strategy?

It should be compatible because Protocol is just special form of ABC class. I will verify that.

**Comment by slmnsh** [comment]
If you don't like the overloads, maybe we can add type hints into separate stub file (.pyi).

**Comment by martindurant** [comment]
> separate stub file 

That's a great idea!
